src/Renci.SshNet/SftpClient.cs

```diff
@@ -2074,60 +2074,68 @@ private IEnumerable<FileInfo> InternalSynchronizeDirectories(string sourcePath,
 
             var sourceDirectory = new DirectoryInfo(sourcePath);
 
-            var sourceFiles = FileSystemAbstraction.EnumerateFiles(sourceDirectory, searchPattern).ToList();
-            if (sourceFiles.Count == 0)
-                return uploadedFiles;
-
-            #region Existing Files at The Destination
-
-            var destFiles = InternalListDirectory(destinationPath, null);
-            var destDict = new Dictionary<string, ISftpFile>();
-            foreach (var destFile in destFiles)
+            using (var sourceFiles = sourceDirectory.EnumerateFiles(searchPattern).GetEnumerator())
             {
-                if (destFile.IsDirectory)
-                    continue;
-                destDict.Add(destFile.Name, destFile);
-            }
+                if(!sourceFiles.MoveNext())
+                    return uploadedFiles;
 
-            #endregion
-
-            #region Upload the difference
-
-            const Flags uploadFlag = Flags.Write | Flags.Truncate | Flags.CreateNewOrOpen;
-            foreach (var localFile in sourceFiles)
-            {
-                var isDifferent = !destDict.ContainsKey(localFile.Name);
+                #region Existing Files at The Destination
 
-                if (!isDifferent)
+                var destFiles = InternalListDirectory(destinationPath, null);
+                var destDict = new Dictionary<string, ISftpFile>();
+                foreach (var destFile in destFiles)
                 {
-                    var temp = destDict[localFile.Name];
-                    //  TODO:   Use md5 to detect a difference
-                    //ltang: File exists at the destination => Using filesize to detect the difference
-                    isDifferent = localFile.Length != temp.Length;
+                    if (destFile.IsDirectory)
+                        continue;
+                    destDict.Add(destFile.Name, destFile);
                 }
 
-                if (isDifferent)
+                #endregion
+
+                #region Upload the difference
+
+                const Flags uploadFlag = Flags.Write | Flags.Truncate | Flags.CreateNewOrOpen;
+                do
                 {
-                    var remoteFileName = string.Format(CultureInfo.InvariantCulture, @"{0}/{1}", destinationPath, localFile.Name);
-                    try
+                    var localFile = sourceFiles.Current;
+                    if (localFile == null)
```

Comment:
`MoveNext()` - returns true if the enumerator was successfully advanced to the next element; false if the enumerator has passed the end of the collection.
`Current` - returns the element in the collection at the current position of the enumerator.

Reply:
I think I knew that already :p
The question is why do you check for **Current** being **null**? You already verified that `MoveNext()` returned **true**, so why the extra null check?

Reply:
`Current` can be null. You know the list can contain the null value. Of course, this is unlikely at this pleace, and I want to avoid hinting about the possibility of null.
```
	public static void Main()
	{
		List<string> list = new List<string>();
		list.Add("A");
		list.Add("B");
		list.Add(null);
		list.Add("D");
		
		using (var e = list.GetEnumerator())
		{
			// check first
			if(!e.MoveNext())
			{
			  return;
			}
			// do some stuff, then enumerate the list
			do {
				Console.WriteLine(e.Current ?? "This is null");
			} 
			while(e.MoveNext()); // stop when we don't have anything else
		}
	}
```

returns:
```
A
B
This is null
D
```

Reply:
Indeed, but I was talking about this specific use. `DirectoryInfo.EnumerateFiles(...)` does not return **null** entries.
I'll remove the null check in a follow-up PR.